graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/TuplePrototypeBuiltins.java

```diff
@@ -0,0 +1,995 @@
+            case forEach:
+                return JSArrayForEachNodeGen.create(context, builtin, args().withThis().fixedArgs(2).createArgumentNodes(context));
+            case keys:
+                return JSTupleIteratorNodeGen.create(context, builtin, JSRuntime.ITERATION_KIND_KEY, args().withThis().createArgumentNodes(context));
+            case map:
+                return JSTupleMapNodeGen.create(context, builtin, args().withThis().fixedArgs(2).createArgumentNodes(context));
+            case reduce:
+                return JSArrayReduceNodeGen.create(context, builtin, false, true, args().withThis().fixedArgs(1).varArgs().createArgumentNodes(context));
+            case reduceRight:
+                return JSArrayReduceNodeGen.create(context, builtin, false, false, args().withThis().fixedArgs(1).varArgs().createArgumentNodes(context));
+            case some:
+                return JSArraySomeNodeGen.create(context, builtin, false, args().withThis().fixedArgs(2).createArgumentNodes(context));
+            case unshifted:
+                return JSTupleUnshiftedNodeGen.create(context, builtin, args().withThis().varArgs().createArgumentNodes(context));
+            case toLocaleString:
+                return JSArrayToLocaleStringNodeGen.create(context, builtin, false, args().withThis().createArgumentNodes(context));
+            case toString:
+                return JSTupleToStringNodeGen.create(context, builtin, args().withThis().fixedArgs(1).createArgumentNodes(context));
+            case values:
+                return JSTupleIteratorNodeGen.create(context, builtin, JSRuntime.ITERATION_KIND_VALUE, args().withThis().createArgumentNodes(context));
+            case with:
+                return JSTupleWithNodeGen.create(context, builtin, args().withThis().fixedArgs(2).createArgumentNodes(context));
+        }
+        return null;
+    }
+
+    public abstract static class JSTupleLengthGetterNode extends JSBuiltinNode {
+
+        public JSTupleLengthGetterNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected long doTuple(Tuple thisObj) {
+            return thisObj.getArraySize();
+        }
+
+        @Specialization(guards = {"isJSTuple(thisObj)"})
+        protected long doJSTuple(DynamicObject thisObj) {
+            return JSTuple.valueOf(thisObj).getArraySize();
+        }
+
+        @Fallback
+        protected long doOther(Object thisObj) {
+            throw Errors.createTypeErrorIncompatibleReceiver(thisObj);
+        }
+    }
+
+    public abstract static class JSTupleToStringNode extends JSBuiltinNode {
+
+        public JSTupleToStringNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected String toString(Tuple thisObj) {
+            return thisObj.toString();
```

Comment:
This call with require an `@TruffleBoundary`. Put that annotation on the outer method (before or after the `@Specialization`) to mark that you don't want PE to look into it. Nothing to optimize here for it, and it should not pull in all the code that might be in the `tuple.toString()` (I didn't look into your implementation there yet, you might have a boundary there or a super-simple implementation that can be PEed ...)

Reply:
I refactored the use of `Tuple.toString()` as calling the object's `join` property is more strictly according to spec 
(BUT I also expect some proposal changes here and in JSTupleToStringNode, let's see).